nf_core/lint_utils.py

```diff
@@ -60,12 +64,45 @@ def run_prettier_on_file(file):
         If Prettier is not installed, a warning is logged.
     """
 
-    try:
-        subprocess.run(
-            ["prettier", "--write", file],
-            stdout=subprocess.DEVNULL,
-            stderr=subprocess.DEVNULL,
-            check=False,
+    _prettier_installed = not shutil.which("prettier") is None
+    _pre_commit_installed = not shutil.which("pre-commit") is None
+
+    if _prettier_installed:
+        _run_prettier_on_file(file)
+    elif _pre_commit_installed:
+        _run_pre_commit_prettier_on_file(file)
+    else:
+        log.warning(
+            "Neither Prettier nor the prettier pre-commit hook are available. At least one of them is required."
         )
-    except FileNotFoundError:
-        log.warning("Prettier is not installed. Please install it and run it on the pipeline to fix linting issues.")
+
+
+def _run_prettier_on_file(file):
+    """Run natively installed Prettier on a file."""
+
+    subprocess.run(
+        ["prettier", "--write", file],
+        stdout=subprocess.DEVNULL,
+        stderr=subprocess.DEVNULL,
+        check=False,
+    )
+
+
+def _run_pre_commit_prettier_on_file(file):
+    """Runs pre-commit hook prettier on a file if pre-commit is installed.
+
+    Args:
+        file (Path | str): A file identifier as a string or pathlib.Path.
+
+    Warns:
+        If Prettier is not installed, a warning is logged.
+    """
+
+    nf_core_pre_commit_config = Path(nf_core.__file__).parent.parent / ".pre-commit-config.yaml"
+
+    subprocess.run(
+        ["pre-commit", "run", f"--config {nf_core_pre_commit_config}", "prettier", f"--files {file}"],
+        stdout=subprocess.DEVNULL,
+        stderr=subprocess.DEVNULL,
+        check=False,
+    )
```

Comment:
Now that we are pre-checking whether the command exists and logging when it doesn't *before* attempting to run, I think we should be using `check=True` and capturing `stdour/stderr` to log any issues when it tries to run `prettier`. 

Currently under this code you could have `prettier` available, so get no warning, it could run but fail for some other reason (who knows, but these things happen) and it would fail silently to the user, looking no different to the output if it was successful. 
```suggestion
def _run_prettier_on_file(file):
    """Run natively installed Prettier on a file."""

    try:
        subprocess.run(
            ["prettier", "--write", file],
            capture_output=True,
            check=True,
        )
    except subprocess.CalledProcessError as e:
        log.error('There was an error running Prettier. Try running manually to fix linting errors in the files')
        log.exception('Failing command: %s', e.cmd)


def _run_pre_commit_prettier_on_file(file):
    """Runs pre-commit hook prettier on a file if pre-commit is installed.

    Args:
        file (Path | str): A file identifier as a string or pathlib.Path.

    Warns:
        If Prettier is not installed, a warning is logged.
   
    """

    nf_core_pre_commit_config = Path(nf_core.__file__).parent.parent / ".pre-commit-config.yaml"
    
    try:
        subprocess.run(
            ["pre-commit", "run", f"--config {nf_core_pre_commit_config}", "prettier", f"--files {file}"],
            capture_output=True,
            check=True,
        )
    except subprocess.CalledProcessError as e:
        log.error('There was an error running Prettier. Try running manually to fix linting errors in the files')
        log.exception('Failing command: %s', e.cmd)
```

Reply:
> Previously although it wasn't transparent if the command ran but failed it would at least tell you prettier hadn't run (claiming it wasn't installed) and that you needed to do it manually.

I disagree - errors that Prettier emitted have been ignored before.

The status quo of the code is

```
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            check=False,
```

which is retained in this PR and basically means we don't want to know if it failed or not. Since - at least the ignorance of any prettier output was added explicitly - and I don't know why - I am reluctant to change this in this PR.

@ErikDanielsson introduced the redirects of the stdout and stderr in https://github.com/nf-core/tools/commit/c8a6045a7c491c64a57e553dfc3906fd8bf8dc57
 and I made [`check=False`](https://github.com/nf-core/tools/commit/5557e17acd33032a002beb25a2f9515fa3a24c07) explicit, but it has been `False` by default before.

What you introduce is surfacing when prettier is run but fails for some other reason than not being installed. I think it is useful, but would like to know what @ErikDanielsson thinks.

Reply:
OK I see what you mean here after re-reading the original code (I didn't fully think through the `FileNotFoundError` usage).

I remain of the opinion it is confusing to run prettier and not report errors unless there is a good reason not to.

Reply:
@awgymer Can you quickly elaborate on why you used `log.exception` in the proposed code? It is not yet used anywhere else in this codebase.

Reply:
Force of habit mainly. In this situation I think the assumption is we aren't likely to be able to predict the sort of errors caused by Prettier but actually I think since it's a subprocess `.exception` probably *isn't* going to be that useful and it would be better to use `e.stdout, e.stderr, e.cmd, e.returncode` in some combination to create an informative error log?

Reply:
It's a little hard to test because I'm not sure the best way to trigger some `prettier` error that isn't it simply not existing?

Reply:
Resolving because I think I included all suggestions and more in the recent changes. Feel free to unresolve or start a new review should I have missed something.